p2p/p2p.go

```diff
@@ -1,61 +1,186 @@
+// Package p2p defines the P2P task, which maintains the network information for the Darknodes. The task pings the
+// Bootstrap nodes upon receiving a `Tick` message and subsequently updates the multi-address store.
 package p2p
 
 import (
+	"encoding/json"
 	"fmt"
+	"math/rand"
+	"net"
+	"time"
 
+	"github.com/renproject/lightnode/rpc"
+	jrpc "github.com/renproject/lightnode/rpc/jsonrpc"
 	"github.com/renproject/lightnode/store"
+	"github.com/republicprotocol/co-go"
+	"github.com/republicprotocol/darknode-go/server/jsonrpc"
 	"github.com/republicprotocol/renp2p-go/core/peer"
-	"github.com/republicprotocol/renp2p-go/foundation/addr"
 	"github.com/republicprotocol/tau"
+	"github.com/sirupsen/logrus"
 )
 
+// P2P handles the peer-to-peer network of nodes.
 type P2P struct {
-	store store.KVStore
+	timeout        time.Duration
+	bootstrapAddrs []peer.MultiAddr
+	logger         logrus.FieldLogger
+	store          store.KVStore
 }
 
-func NewP2P(store store.KVStore) *P2P {
-	return &P2P{
-		store: store,
+// New returns a new P2P task.
+func New(logger logrus.FieldLogger, cap int, timeout time.Duration, store store.KVStore, bootstrapAddrs []peer.MultiAddr) tau.Task {
+	p2p := &P2P{
+		timeout:        timeout,
+		logger:         logger,
+		store:          store,
+		bootstrapAddrs: bootstrapAddrs,
 	}
+	return tau.New(tau.NewIO(cap), p2p)
 }
 
+// Reduce implements the `tau.Task` interface.
 func (p2p *P2P) Reduce(message tau.Message) tau.Message {
 	switch message := message.(type) {
-	case QueryAddressRequest:
-		return p2p.handleQueryAddress(message)
+	case Tick:
+		return p2p.handleTick(message)
+	case rpc.QueryMessage:
+		return p2p.handleQuery(message)
 	default:
-		panic(fmt.Errorf("unexpected message type %T", message))
+		p2p.logger.Panicf("unexpected message type %T", message)
 	}
+	return nil
 }
 
-func (p2p *P2P) handleQueryAddress(message QueryAddressRequest) tau.Message {
-	var multi peer.MultiAddr
-	err := p2p.store.Read(message.Addr.String(), &multi)
-	if err != nil {
-		return tau.NewError(err)
-	}
-	return QueryAddressResponse{
-		Multi: multi,
+// handleTick queries the Bootstrap nodes for their peers. Upon receiving a response, we update the store with the
+// multi-address of the node we queried, as well as any nodes it returns. If we do not receive a response, we remove it
+// from the store if it previously existed.
+func (p2p *P2P) handleTick(message tau.Message) tau.Message {
+	request := jsonrpc.JSONRequest{
+		JSONRPC: "2.0",
+		Method:  jsonrpc.MethodQueryPeers,
+		ID:      rand.Int31(),
 	}
-}
 
-func New(cap int, store store.KVStore) tau.Task {
-	p2p := NewP2P(store)
-	return tau.New(tau.NewIO(cap), p2p)
+	go func() {
```

Comment:
Given that you create a background goroutine with multiple concurrent workers that make updates to a mutex-protected store, why bother with the `Tick` message at all?

On creation of the P2P task, spawn a background goroutine similar to this one that loops (and each loop waits some number of seconds/minutes given as an argument to the creation of the P2P task). This completely encapsulates the background updates being made since parent tasks do not need to drive the P2P task with tick messages.